instrumentation/elasticsearch/elasticsearch-rest-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/elasticsearch/rest/ElasticsearchEndpointDefinition.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.elasticsearch.rest;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+public class ElasticsearchEndpointDefinition {
+
+  private static final String UNDERSCORE_REPLACEMENT = "0";
+
+  private final String endpointName;
+  private final List<Route> routes;
+
+  private final boolean isSearchEndpoint;
+
+  public ElasticsearchEndpointDefinition(
+      String endpointName, String[] routes, boolean isSearchEndpoint) {
+    this.endpointName = endpointName;
+    this.routes = Arrays.stream(routes).map(Route::new).collect(Collectors.toList());
+    this.isSearchEndpoint = isSearchEndpoint;
+  }
+
+  @Nullable
+  public String getEndpointName() {
+    return endpointName;
+  }
+
+  public boolean isSearchEndpoint() {
+    return isSearchEndpoint;
+  }
+
+  public void processPathParts(String urlPath, BiConsumer<String, String> consumer) {
+    for (Route route : routes) {
+      if (route.hasParameters()) {
+        Matcher matcher = route.createMatcher(urlPath);
+        if (matcher.find()) {
+          for (String key : route.getPathPartNames()) {
+            String value = matcher.group(key);
+            if (key.contains(UNDERSCORE_REPLACEMENT)) {
+              // replace underscore back
+              key = key.replace(UNDERSCORE_REPLACEMENT, "_");
+            }
+            consumer.accept(key, value);
+          }
+          return;
+        }
+      }
+    }
+  }
+
+  List<Route> getRoutes() {
+    return routes;
+  }
+
+  static final class Route {
+    private final String name;
+    private final boolean hasParameters;
+    private volatile EndpointPattern epPattern;
+
+    public Route(String name) {
+      this.name = name;
+      this.hasParameters = name.contains("{") && name.contains("}");
+    }
+
+    String getName() {
+      return name;
+    }
+
+    boolean hasParameters() {
+      return hasParameters;
+    }
+
+    List<String> getPathPartNames() {
+      return getEndpointPattern().getPathPartNames();
+    }
+
+    Matcher createMatcher(String urlPath) {
+      return getEndpointPattern().getPattern().matcher(urlPath);
+    }
+
+    private EndpointPattern getEndpointPattern() {
+      if (epPattern == null) {
+        epPattern = new EndpointPattern(this);
+      }
+
+      return epPattern;
```

Comment:
In general this is not a thread safe pattern of setting a volatile field. In this particular case I think we're fine if that, it's just caching the pattern; it still deserves a comment saying that it's intentional and fine.

Reply:
Agree! It is intentional "`volatile`-only" instead of synchronizing here. Worst case is that it will be created more than once. Main purpose is to cache the pattern without paying the cost of synchronizing. Will add a comment.